### ISSUE TITLE
BOCFEL: Trinity skips some text when entering the white door

### ISSUE THREAD

**Issue by angstsmurf** [garglk/garglk#797]
From the current  [let's play of _Trinity_](https://intfiction.org/t/lets-play-trinity-by-brian-moriarty/65799/26):

> The following text didn’t display in the game for me, so I had to pull it from the transcript:
> 
> > >e
> > As you wade to the threshold a familiar roadrunner flutters past. The ruby in its beak gleams enticingly as it slips through the white door.
> > 
> > All color abruptly drains from the landscape. Trees, sky and sun flatten into a spherical shell, with you at the very center. A hissing in your ears becomes a rumble, then a roar as the walls of the shell collapse inward, faster and faster.
> > 
> > “This way, please.”
> > 
> > You turn, but see no one.
> > 
> > “This way,” the voice urges. “Be quick.”
> > 
> > The space around you articulates. “No!” your mind shudders. “That’s not a direction!”
> > 
> > “It’s a perfectly legitimate direction,” retorts the voice with cold amusement. “Now come along.”

I haven't checked the source code, so I'm not sure what exactly the game is doing here. It does not happen in the original Infocom interpreter (EDIT: or in Frotz. They both print the text followed by a [More] prompt, and then the title page). Attaching a zipped save file for release 12 / 860926. Just type >E.

[Trinity.glksave.zip](https://github.com/garglk/garglk/files/13533504/Trinity.glksave.zip)

EDIT: Now I have checked the ZIL code. [See comment below.](https://github.com/garglk/garglk/issues/797#issuecomment-1837230664)

**Comment by angstsmurf** [comment]
The relevant code seems to be[ about here](https://github.com/the-infocom-files/trinity/blob/7eb27f9a7cc429fb79d1bd1032369939afeec946/things.zil#L3557-L3568):
```
	 <CARRIAGE-RETURNS>
	 <CLEAR -1>
```

where CARRIAGE-RETURNS seems to [simply print 23 line feeds](https://github.com/the-infocom-files/trinity/blob/7eb27f9a7cc429fb79d1bd1032369939afeec946/misc.zil#L1529-L1537):
```
<ROUTINE CARRIAGE-RETURNS ("OPTIONAL" (CNT 23))
	 <REPEAT ()
		 <COND (<ZERO? .CNT>
			<RTRUE>)>
		 <CRLF>
		 <DEC CNT>>
	 <RTRUE>>
```
So, my understanding is that the game expects the interpreter to halt with a "[More]" prompt if enough carriage returns are printed, the screen fully updated, and not proceed until a key is pressed. All the GLK based interpreters I've tried seem to instead carry on processing the CLEAR -1 instruction immediately, clearing all the text just printed.

**Comment by angstsmurf** [comment]
As mentioned by @eriktorbjorn in the forum topic linked above, there is similar code in [Beyond Zork](https://github.com/the-infocom-files/beyondzork/blob/3f707f00a0cf9ceaff5632a465968f5644b81d77/rarities.zil#L2147C1-L2158C13):

```
<ROUTINE ENDING ("AUX" X LEVEL)
	 <TELL
"A devastating ground shock sends you sprawling! The roof of the cavern gives way at the same moment, and you watch helplessly as tons of granite crumble all around you..">
	 <SET X <LOC ,COCO>>
	 <COND (<OR <ZERO? .X>
		    <AND <NOT <EQUAL? .X ,PLAYER>>
			 <NOT <IN? .X ,PLAYER>>>>
		<JIGS-UP>
		<RFALSE>)>
	 <TELL ,PERIOD>
	 <CARRIAGE-RETURNS>
	 <CLEAR -1>
```

**Comment by cspiegel** [comment]
This is indeed a sticky problem.

A patch is a possible solution, if slightly less than ideal. The assembly looks like this:

```
2d487:  88 44 30 00             CALL_1S         110c0 -> -(SP)
2d48b:  ed 3f ff ff             ERASE_WINDOW    #ffff
```
You can't remove the call to `@erase_window`, so you're left with 4 bytes.  This can be converted to `@read_char 1 -> sp`, and you can do this in Bocfel's config file, like this:

```
# Trinity
[12-860926]
patch = 2d487 4 [88 44 30 00] [f6 7f 01 00]
```
This shows the text and waits for input, but there's no indication that you should hit a key to continue, although everybody would figure it out quickly enough. This may be a clean enough solution.

There are 13 other calls to `CARRIAGE-RETURNS` though none seem to be directly followed by clearing the screen, so it might be enough to just do this patch.  And, of course, a similar one for Beyond Zork

**Comment by angstsmurf** [comment]
That seems like a very clean and simple solution.

**Comment by cspiegel** [comment]
Unfortunately, Beyond Zork uses `@call_1n` to call `CARRIAGE-RETURNS`, which is 3 bytes, so the simple fix here doesn't work.

However, the code looks like this:

```
3ca37:  8f 44 55                CALL_1N         11154
3ca3a:  ed 3f ff ff             ERASE_WINDOW    #ffff
3ca3e:  7b 13 83                SET_COLOUR      G03,G73
3ca41:  bb                      NEW_LINE        
3ca42:  b2 ...                  PRINT           ""Is "
```
Given that `@erase_window -1` clears the screen, that call to `@new_line` probably isn't too critical, so the extra byte can be taken from there.  As such, here's a proposed patch:

```
[57-871221]
patch = 0x3ca37 11 [8f 44 55 ed 3f ff ff 7b 13 83 bb] [f6 7f 01 00 ed 3f ff ff 7b 13 83]
```

Resulting in:

```
3ca37:  f6 7f 01 00             READ_CHAR       #01 -> -(SP)
3ca3b:  ed 3f ff ff             ERASE_WINDOW    #ffff
3ca3f:  7b 13 83                SET_COLOUR      G03,G73
3ca42:  b2 ...                  PRINT           ""Is "
```
But I don't know Beyond Zork well enough to get to this point to test it out.

If somebody's familiar with BZ and wants to help out, this bit comes right after the text:

> A devastating ground shock sends you sprawling! The roof of the cavern gives way at the same moment, and you watch helplessly as tons of granite crumble all around you..

You can either directly try the patch, or provide a save file right before this occurs, which I can use to test.

**Comment by angstsmurf** [comment]
Here is a save file for Beyond Zork release 60 / 880610. Just type WAIT or anything that passes time.

[beyondzork-end.glksave.zip](https://github.com/garglk/garglk/files/13574915/beyondzork-end.glksave.zip)

**Comment by angstsmurf** [comment]
Excellent! Thanks.

**Comment by cspiegel** [comment]
OK, I've release Bocfel 2.1.2 with this fix, and it's now merged in #798.

**Comment by cspiegel** [comment]
> Excellent! Thanks.
> 
> EDIT: Not sure if I did something wrong when applying the patch, but I had to add `&` (address of) prefixes to the Trinity serial number strings in order to make it compile, but after that it worked just fine.

Sorry, when making that patch it was against master, not 2.1.1, and I'd forgotten that I changed the data structure of the patches slightly.  It used to be a pointer to char array (`char (*)[7]`) but I changed it to a reference (`char (&)[7]`) to be more ergonomic, i.e. to not require the `&`.  Just adding it back is the right fix (it's what I did when cherry picking the patches to my 2.1 branch).
